src/ParametersHandler/YarpImplementation/src/YarpImplementation.cpp

```diff
@@ -186,6 +186,17 @@ std::string YarpImplementation::toString() const
     return output;
 }
 
+std::unordered_set<std::string> YarpImplementation::getParametersNames() const
+{
+    std::unordered_set<std::string> names;
+    for (const auto& [name, _] : m_lists)
+    {
+        names.insert(name);
+    }
```

Comment:
Notice that this does not return the parameters, but only the name of the subgroups. In order to get the name of the parameters, you may need to scan the ``Bottle`` as in https://github.com/dic-iit/bipedal-locomotion-framework/blob/be1e0a2ea0077e3ed745ec6891aac3e71d35824a/src/ParametersHandler/YarpImplementation/src/YarpImplementation.cpp#L122-L144

Reply:
OMG every time I see this Bottle-related code my hands get glued on my hairs for a while

Reply:
Is there any reason why the storage of the data is based on YARP types? I understand the need to parse _from_ a YARP object, but the underlying storage of the YARP backend could be exactly the same of e.g. the Std backend, and maybe some code could be made implementation-independent. If there is no specific purpose, using YARP here just breaks any kind of readability :)

After the random rant (sorry) it's not clear to me what I should get here. Is it allowed to have the same name for two different parameters belonging to different groups? Is it allowed to have a group named as a parameter?

To improve clarity regards the above questions, maybe having two different helpers like `getParameterNames` and `getGroupNames` would be more intuitive. I'll go towards this direction if there are no objections, assuming that groups and parameters cannot clash.

---

Then, I found a bug in the `StdImplementation`, I write it here, let me know if I need to open an issue. The following always returns an expired weak pointer:

https://github.com/dic-iit/bipedal-locomotion-framework/blob/be1e0a2ea0077e3ed745ec6891aac3e71d35824a/src/ParametersHandler/src/StdImplementation.cpp#L127-L131

Reply:
New tentative: 5eeb3ef. Please double check the YARP implementation, it's not yet clear to me how data is stored in the Bottle.

Reply:
>Is there any reason why the storage of the data is based on YARP types? I understand the need to parse from a YARP object, but the underlying storage of the YARP backend could be exactly the same of e.g. the Std backend, and maybe some code could be made implementation-independent. If there is no specific purpose, using YARP here just breaks any kind of readability :)

Yes. The simple reason is the necessity of reading ``yarp`` configuration files. They get in as ``Searchable`` and the conversion to a ``std::any`` (as in the std implementation) is not possible since the type of the parameter is not known beforehand.

>After the random rant (sorry) it's not clear to me what I should get here. Is it allowed to have the same name for two different parameters belonging to different groups? Is it allowed to have a group named as a parameter?

It does not throw an error, but in the ``yarp`` implementation the ``group`` has the precedence, while in the ``std`` implementation only the last added survive (since it uses a ``map``). So I would say it is a corner case that is better to avoid. Two parameters belonging to the same group are totally allowed.

> To improve clarity regards the above questions, maybe having two different helpers like getParameterNames and getGroupNames would be more intuitive. I'll go towards this direction if there are no objections, assuming that groups and parameters cannot clash.

I am not sure whether this is possible in the ``std`` implementation since the handler does not know if a specific entry is a parameter or a group. It is up to the user to call the correct methods. I would suggest to eventually stop at the first level, i.e. print all parameter and group names, but not the parameters inside the subgroups.

> Then, I found a bug in the StdImplementation, I write it here, let me know if I need to open an issue. The following always returns an expired weak pointer:

It is not a bug. If the group does not exist it returns a nullptr.

Anyway, this for ``yarp`` may work:
```
     for (size_t i = 0; i < m_container.size(); i++) 
     { 
         yarp::os::Value& bb = bot.get(i); 
  
         yarp::os::Bottle* sub = bb.asList(); 
         if ((sub) && (sub->size() >= 1)) 
         { 
             std::string name = sub->get(0).toString(); 
             names.insert(name); 
  
         } else //This is a bit weird case where the parameter does not have a value, so probably you can just add it as it is
         { 
             names.insert(bb.toString()); 
         } 
     } 
```

Reply:
>New tentative: 5eeb3ef. Please double check the YARP implementation, it's not yet clear to me how data is stored in the Bottle.

I saw this later, please check https://github.com/dic-iit/bipedal-locomotion-framework/pull/134#discussion_r518591209

Reply:
> Yes. The simple reason is the necessity of reading yarp configuration files. They get in as Searchable and the conversion to a std::any (as in the std implementation) is not possible since the type of the parameter is not known beforehand.

Ahn now I see! This is a big limitation, I have never been a big fan of this approach but I think I have to live with that. 

> Two parameters with the same name belonging to different groups are totally allowed.

This is a big problem as well. I propose the following:

1. `getParameterNames` returns all the parameters stored in the current `IParametersHandler` without considering subgroups
1. `getGroupNames` returns all the group names

Then, if you want to know all the parameters of an object (that can have duplicates due to subgroup handling) users have to manually loop over groups.

The alternative is providing a _scoped name_ of the parameters like `group_name::param_name`, but this would go beyond the scope of this PR.

> I am not sure whether this is possible in the std implementation since the handler does not know if a specific entry is a parameter or a group. It is up to the user to call the correct methods. I would suggest to eventually stop at the first level, i.e. print all parameter and group names, but not the parameters inside the subgroups.

Ok I'm reading your comment sequentially and I now realized that you suggested the same thing :)

> It is not a bug. If the group does not exist it returns a nullptr.

Maybe returning directly `nullptr` would be more clear :) I'm pretty sure that who wrote those lines wanted to do something else xD

> Anyway, this for `yarp` may work

Will try, thanks!

Reply:
>Then, if you want to know all the parameters of an object (that can have duplicates due to subgroup handling) users have to manually loop over groups.

Agree. Indeed, a subgroup is yet another ``IParametersHandler`` object.

> Maybe returning directly nullptr would be more clear :) I'm pretty sure that who wrote those lines wanted to do something else xD

The important thing is to return an expired ``weak_ptr``. This is one of the possible ways. It was me proposing the ``weak_ptr`` mechanism in the first place :grin:

Reply:
Well, it took me a bit more than expected and more LOC than I wanted, but with 4865a32 the `YarpImplementation` should be ok. I had to change a bit the logic because otherwise it was impossible to separate groups from vector parameters.

Reply:
As per https://github.com/dic-iit/bipedal-locomotion-framework/pull/134#issuecomment-725388389, the `IParametersHandler` modification have been removed from this PR.